Updates to scRNA-seq normalization notebook (#251)
AlexsLemonade/training-modules: scRNA-seq/02-normalizing_scRNA-seq.Rmd

## Set up a SingleCellExperiment object from count data

Now that we have filtered our data, we are ready to normalize it.
To do this, we are going to use some single-cell RNA-seq R packages called `scater` and `scran`. 

![](diagrams/full-length_2.png)

The functions in these packages require the data to be in their own special object type (not an uncommon thing for R packages to do) called `SingleCellExperiment`.
So first, we are going to set up our data in this format.
To learn more about `SingleCellExperiment` objects and how it works with Bioconductor packages, we recommend this article: [Amezquita *et al.* 2020.](https://www.nature.com/articles/s41592-019-0654-x).

```{r}
# Set up the Single Cell Experiment object 
gbm_sce <- SingleCellExperiment(list(counts = as.matrix(sc_data_filtered[, -1])))

# Store the gene names in this object
rownames(gbm_sce) <-  sc_data_filtered$gene
```

> In the intro-to-R-tidyverse module notebook, `01-intro-to-r`, we discuss base R object types, but there are some 'special' object types that are package-specific. 

Reviewer: ```suggestion
In the intro-to-R-tidyverse module notebook, [`01-intro-to-base_R.Rmd`](https://htmlpreview.github.io/?https://github.com/AlexsLemonade/training-modules/blob/master/intro-to-R-tidyverse/01-intro_to_base_R.nb.html) we discuss base R object types, but there are some 'special' object types that are package-specific. 
```

Author: I don't love linking this because it may break for updated versions, and I can't think of a way to properly handle that.

Reviewer: Meh. Delete it then. But probably make sure the notebook name is accurate. aka keep the `_base_R.Rmd` part.

Author: yup, kept that!

# R packages, explore its properties here. 
# Also check it out in the environment tab.
gbm_sce
```

Below is a figure from [Amezquita *et al.* (_bioRxiv._ 2019)](https://www.biorxiv.org/content/10.1101/590562v1) (the preprint version of the above-linked paper) that shows the general structure of `SingleCellExperiment` objects. 

![](figures/sce_structure.png)

Note that each function which transforms the `SingleCellExperiment` object can store its results back into the object (or a new copy of the object, but that tends to be a waste of space).
Each of these transformation functions also have a corresponding extraction function that allow you extract (or set) those results from the `SingleCellExperiment` object. 

For more information on `SingleCellExperiment` objects, check out this [excellent vignette](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)
and/or [Amezquita *et al.* (2020)](https://www.nature.com/articles/s41592-019-0654-x)

### Adding more metadata to the SCE object

```{r sample_info}
# get the sample names and put them in a data frame (1 column)
sample_info <- tibble::tibble(geo_accession = colnames(gbm_sce)) %>%

>   #join to the metadata

Reviewer: ```suggestion
  # join to the metadata
```

  dplyr::left_join(sc_metadata) %>%
  dplyr::select(geo_accession,
                title,
                cell_type = cell.type.ch1,
                tissue = characteristics_ch1.3
                ) %>%
  # clean up tissue
  dplyr::mutate(tissue = stringr::str_remove(tissue, "tissue: "))
```

Check that the sample info accession ids are the same as the columns of our data

```{r check_sampleinfo, live = TRUE}
all.equal(sample_info$geo_accession, colnames(gbm_sce))
```

Now add that data to the SCE object.
For unclear reasons, we have to convert our table to a `DataFrame` object in order for this to work.

> Just to keep things confusing, a `DataFrame` is not the same as a `data.frame` that we have been using throughout. 

Author: Me too.

We also need to be sure to include the `row.names` argument to keep those properly attached

```{r}
colData(gbm_sce) <- DataFrame(sample_info, row.names = sample_info$geo_accession)
```

## Normalization of count data 

In whatever data we are working with, we are always looking to maximize biological variance and minimize technical variance. 
A primary source of technical variation we are concerned with is the variation in library sizes among our samples. 

> While different cells may have different total transcript counts, it seems far more likely that the primary source of variation that we see is due to library construction, amplification, and sequencing. 

Author: I will delete "far".

> While different cells may have different total transcript counts, it seems far more likely that the primary source of variation that we see is due to library construction, amplification, and sequencing. 

Reviewer: Deal! 🤝

This is where normalization methods usually come into the workflow.
The distribution of the counts that we saw in the previous notebook, and in particular the fact that the count data is noisy with many zero counts, makes normalization particularly tricky.
To handle this noise, we normalize cells in groups with other cells like them; a method introduced in [Lun *et al.* (2016)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7).

Briefly, we first cluster the cells to find groups of similar cells, then compute normalization factors based on the sums of expression in those groups. 
The group normalization is then applied back to the individual cells within the group to create a normalized count matrix. 
In this case, we will also log-transform the normalized counts to get a less skewed distribution of expression measures. 

> Note that because of the zero counts, the logNormCounts() function will add a pseudocount of 1 to each value before computing the log.

Reviewer: ```suggestion
Note that because of the zero counts, the `logNormCounts()` function will add a pseudocount of 1 to each value before computing the log.
```

```{r pca}
# Use PCA for dimension reduction of cells' scran normalized data
norm_pca <- scater::calculatePCA(gbm_sce, ntop = nrow(gbm_sce)) 

# PCA on the raw counts, log transformed
log_pca <- counts(gbm_sce) %>% # get the raw counts
  log1p() %>% # log transform to make these more comparable to the normalized values
  scater::calculatePCA(ntop = nrow(gbm_sce)) # calculate PCA scores

```

Note that we are using `scater::calculatePCA()` two different ways here: once on the full `gbm_sce` object, and once on just the `counts` matrix. 
When we use `calculatePCA()` on the object, it automatically uses the log normalized matrix from inside the object.

We are also using the `ntop` argument so that the PCA is calculated using data from all genes in the data set. 
This is not the default for `calculatePCA()` as we will explore below.

Next we will arrange the PCA scores for plotting, adding a column with the cell type data so we can color each point of the plot. 

```{r pca_df}
# Set up the PCA scores for plotting
norm_pca_scores <- data.frame(norm_pca,
                              geo_accession = rownames(norm_pca),
                              cell_type = gbm_sce$cell_type)

> log_pca_scores <- data.frame(log_pca,
>                               geo_accession = rownames(log_pca),
>                               cell_type = gbm_sce$cell_type)

Reviewer 2: ```suggestion
log_pca_scores <- data.frame(log_pca,
                             geo_accession = rownames(log_pca),
                             cell_type = gbm_sce$cell_type)
```

  geom_point(alpha = 0.05) +
  stat_function(fun = metadata(gbm_variance)$trend, color = "blue") + 
  xlab("Mean log-expression") + 
  ylab("Variance")

```

```{r get_highvar}
highvar_genes <- scran::getTopHVGs(gbm_variance, n = 500)
highvar_pca <- scater::calculatePCA(gbm_sce, subset_row = highvar_genes)
```

```{r highvar_df}
highvar_pca_scores <- data.frame(highvar_pca,
                                 geo_accession = rownames(highvar_pca),
                                 cell_type = gbm_sce$cell_type)
```

```{r plot_highvar}
ggplot(highvar_pca_scores, aes(x = PC1, y = PC2, color = cell_type)) +
  geom_point() + 
  ggtitle("High variance gene PCA scores") + 
  colorblindr::scale_color_OkabeIto()
```

Now what we didn't tell you before, but will tell you now, is that the default for `calculatePCA()` is to use the top 500 highest variance genes. 
This is what we were overriding when we used `ntop = nrow(gbm_sce)`. 
So if we want to shortcut what we did to get the above plot (or one very close to it), we could have done it with `scater::calculatePCA(gbm_sce)` using no additional arguments.

*Why didn't we do that first?*

> Well, it wouldn't have shown you the effects of normalization or making the choice to include only high variance genes. 😉

Reviewer 2: 😆

To find genes that are distinct to one cell type or another, we can use the `scran::findMarkers()` function to test for genes that are significantly different among cells of each type.
This will return a list of tables, one for each cell type, with statistics for each gene showing how well it differentiates that cell type against all other types.

```{r find_markers}
markers <- scran::findMarkers(gbm_sce, gbm_sce$cell_type)
```

Next we can look at one of those tables, in this case the one for neurons. 
We have done some transformation here to pull the gene name into a column of its own and filtered to the genes that seem to differentiate best using the `Top` column.

```{r marker_table}
markers$Neuron %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  dplyr::filter(Top <= 100) # most significant 100 for each pairwise comparison

```

Because we tend to like tidy data, here we use a `tidyverse` function from the [`purrr` package](https://purrr.tidyverse.org) to apply the same operations as above to every element of the `markers` list. 

> `purrr::map_df` is particularly nice, because after applying the operations in the second argument, it combines the results into a single data frame, adding a column with the name specified by `.id` to differentiate the rows that came from each element of the list.

Reviewer: ```suggestion
`purrr::map_df()` is particularly nice, because after applying the operations in the second argument, it combines the results into a single data frame, adding a column with the name specified by `.id` to differentiate the rows that came from each element of the list.
```

```{r top_markers}
# Make a data frame of the top markers for each cell type
top_markers <- purrr::map_df(as.list(markers),
                             ~ as.data.frame(.) %>%
                               tibble::rownames_to_column("gene") %>%
                               dplyr::filter(Top <= 100),

>                              .id = "cell_type")

Reviewer 2: Forgot to ask about writing out results for use in pathway analysis module (#243) here!

>                              .id = "cell_type")

Reviewer: I'm worried about completely losing our participants on this `purrr`.

```

This list has a lot of redundancy, and we really only care about the gene names for the next step, so we will reduce that column down to just the unique set of gene names with `unique()`.

```{r unique_markers}
unique_markers <- unique(top_markers$gene)
length(unique_markers)
```

> Now we will use those unique markers to calculate the principal component analysis using just those genes.

Reviewer: ```suggestion
Now we will use those unique markers to calculate the PCA using just those genes.
```

```{r marker_pca}
marker_pca <- calculatePCA(gbm_sce, 
                           subset_row = unique_markers, 
                           ntop = length(unique_markers)) %>% # use all of the markers 
  as.data.frame() %>%
  tibble::rownames_to_column("geo_accession") %>%

>   dplyr::mutate(cell_type = gbm_sce$cell_type)

Author: Yeah, that was some part of my thinking. I don't feel strongly one way or another.

```

Now we can make a plot of the marker gene PCA!

```{r plot_marker, live = TRUE}
ggplot(marker_pca, aes(x = PC1, y = PC2, color = cell_type)) +
  geom_point() + 
  ggtitle("Marker gene PCA scores") + 
  colorblindr::scale_color_OkabeIto()
```

## Save the normalized data to tsv file

In case we wanted to return to this data later, let's save the normalized data
to a tsv file. 
In order to do this we need to extract our normalized counts from `gbm_sce`. 

> Refer back to the `SingleCellExperiment` figure above to determine why we are using this `logcounts` function.

Reviewer: ```suggestion
Refer back to the `SingleCellExperiment` figure above to determine why we are using this `logcounts()` function.
```

The main functions in `purrr` are the `map()` functions, which take as their main arguments a **list** and a **function** to apply to each element of the list. 
The first one, `purrr::map()` is essentially equivilent to the `lapply()` function that is part of base R, but with some different defaults. 
We will use it first to get the top rows from each table by using it to apply the `head()` function to each element of the list.
The results are returned as a new list.

```{r head_markers, eval = FALSE}
purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
           head) # the function to apply (note no parenthesis)
```

Since that would return a list of data frames, it isn't quite what we want.
There is no built-in function that will give us just the first few rownames, so we will have to define one, and `purrr` gives us a nice shorthand for doing that with `~` syntax.

> Each element of the list will be assigned to the `.x` variable, so we can define an expression that uses that and `purrr::map()` will apply that to each element of the list.

Author: I went a different direction revising this. See you you like my new wording: https://github.com/AlexsLemonade/training-modules/blob/afc6daed4372489b301c374ea74e529d64fa8d69/scRNA-seq/02-normalizing_scRNA-seq.Rmd#L333-L335

purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
           ~ rownames(head(.x)) )# our custom function.
```

Another variant is `purrr::imap()`, which allows us to use the names of the list elements in our function. 
We will use that here to name output files where we will print each of the marker tables, one for each cell type.
We are again defining a custom function within the call to `purrr:imap()` using the `~` syntax, but this time we can use `.x` for the list elements and `.y` for their names.

In addition to writing the tables out, we are saving the data frames we created as a new list object, that we can use in the next step.

```{r write_tables}
marker_df_list <- purrr::imap(
  as.list(markers), # convert markers to a 'regular' list for purrr
  ~ as.data.frame(.x) %>% # convert the table to a data frame
    tibble::rownames_to_column("gene") %>% # make genes a new column
    readr::write_tsv( # write each data frame to a file
      file.path(marker_dir, paste0(.y, "_markers.tsv")) # construct the output path
      )
  )
```

`purrr::map_df()` is another nice function in `purrr`. 

> It's special purpose is that after it applies the operations in the second argument, it combines the results into a single data frame.

Author: That was a reference! I will remove it if requested, but I need my little bits of dad humor sometimes.

> It's special purpose is that after it applies the operations in the second argument, it combines the results into a single data frame.

Reviewer: The dad-joke was lost on me ¯\_(ツ)_/¯

```

Below is a figure from [Amezquita *et al.* (_bioRxiv._ 2019)](https://www.biorxiv.org/content/10.1101/590562v1) (the preprint version of the above-linked paper) that shows the general structure of `SingleCellExperiment` objects. 

![](figures/sce_structure.png)

Note that each function which transforms the `SingleCellExperiment` object can store its results back into the object (or a new copy of the object, but that tends to be a waste of space).
Each of these transformation functions also have a corresponding extraction function that allow you extract (or set) those results from the `SingleCellExperiment` object. 

For more information on `SingleCellExperiment` objects, check out this [excellent vignette](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html)
and/or [Amezquita *et al.* (2020)](https://www.nature.com/articles/s41592-019-0654-x)

### Adding more metadata to the SCE object

```{r sample_info}
# get the sample names and put them in a data frame (1 column)
sample_info <- tibble::tibble(geo_accession = colnames(gbm_sce)) %>%
  # join to the metadata
  dplyr::left_join(sc_metadata) %>%
  dplyr::select(geo_accession,

>                 title,
>                 cell_type = cell.type.ch1,
>                 tissue = characteristics_ch1.3

Reviewer 2: ```suggestion
                  title,
                  cell_type = cell.type.ch1,
                  tissue = characteristics_ch1.3
```
Probably?

Author: looks right to me?

```

We've plotted the counts data for you. Knowing that we want the same graph, but different data, use the above template to plot the normalized data. 
Feel free to customize the plot with a different theme or color scheme!

Let's plot the `norm_pca_scores` data:

```{r norm_pca_plot, live = TRUE}
ggplot(norm_pca_scores, aes(x = PC1, y = PC2, color = cell_type)) +
  geom_point() + 
  ggtitle("Normalized log counts PCA scores") + 
  colorblindr::scale_color_OkabeIto()
```

Do you see an effect from the normalization in the comparison between these plots?

## Modeling variance

The variation in gene expression we see among cells comes from a combination of variation due to technical effects and the biology we really care about. 
In order to account for this we could just take the largest variance genes, but if we want to be a bit more careful about it, we can model the variance in expression of each gene as a function of mean expression for that gene.
This is useful because we generally expect the variance to increase as mean expression increases, even if there is no biological signal in the expression variation.

> We will do this modeling of variance by expressionwith the `scran::modelGeneVar()` function, saving the results to a new variable. 

Reviewer 2: ```suggestion
We will do this modeling of variance by expression with the `scran::modelGeneVar()` function, saving the results to a new variable. 
```

Quoting from the docs for `modelGeneVar()` for note to self reasons

> For each gene, we compute the variance and mean of the log-expression values. A trend is fitted to the variance against the mean for all genes using fitTrendVar. The fitted value for each gene is used as a proxy for the technical component of variation for each gene, under the assumption that most genes exhibit a low baseline level of variation that is not biologically interesting. The biological component of variation for each gene is defined as the the residual from the trend.

Author: Yeah, I will admit I don't love this model as a representation of the two components of variance. There is another modelGeneVarPoisson() (or something like), which seems like it might be more reasonable. But I like the point that doing this modeling makes: not all the variation we see is biological, and there is a relationship between mean expression and variation that we should keep ourselves aware of.

```{r model_variance}
gbm_variance <- scran::modelGeneVar(gbm_sce)
```

> Now lets plot the relationship between gene expression and variance we were discussing.

Reviewer 2: ```suggestion
Now let's plot the relationship between gene expression and variance we were discussing.
```

Here we will also add the fitting curve that `scran::modelGeneVar()` created, which is stored as function in the  `$trend` slot of the `gbm_variance` object.

> We can add a function like that curveto a `ggplot` with a `stat_function` layer.

Reviewer 2: ```suggestion
We can add a function like that curve to a `ggplot` with a `stat_function` layer.
```

```{r plot_variance}
ggplot(as.data.frame(gbm_variance), aes(x = mean, y = total)) +
  geom_point(alpha = 0.05) +
  stat_function(fun = metadata(gbm_variance)$trend, color = "blue") + 
  xlab("Mean log-expression") + 
  ylab("Variance")

```

Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and calculate PCA scores using only those genes. 
Here we are picking the 500 top genes.

```{r get_highvar}
highvar_genes <- scran::getTopHVGs(gbm_variance, n = 500)
highvar_pca <- scater::calculatePCA(gbm_sce, subset_row = highvar_genes)
```

```{r highvar_df}
highvar_pca_scores <- data.frame(highvar_pca,
                                 geo_accession = rownames(highvar_pca),
                                 cell_type = gbm_sce$cell_type)
```

```{r plot_highvar}
ggplot(highvar_pca_scores, aes(x = PC1, y = PC2, color = cell_type)) +
  geom_point() + 
  ggtitle("High variance gene PCA scores") + 
  colorblindr::scale_color_OkabeIto()
```

> Now what we didn't tell you before, but will tell you now, is that the default for `calculatePCA()` is to use the top 500 highest variance genes. 

Reviewer 2: Not clear to me from my quick read of documentation if `calculatePCA()` includes the `modelGeneVar()` `getTopHVGs()` steps, but I strongly suspect you've read more about this than me and also that design choice seems super reasonable.

This is what we were overriding when we used `ntop = nrow(gbm_sce)`.

> *However*, that filter is based on total variance, not biological variance, as we have done here. 

Reviewer 2: Oh I see, just needed to read a bit further!

Author: 😄 So should I, because I was starting to write all of this again in reply to your previous comment!

The first one, `purrr::map()` is essentially equivalent to the `lapply()` function that is part of base R, but with some different defaults. 
We will use it to get the top rows from each table by applying the `head()` function to each element of the list.
The results are returned as a new list.

```{r head_markers, eval = FALSE}
purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr
           head) # the function to apply (note no parenthesis)
```

This returns a list of data frames, which isn't quite what we want.

There is no built-in function that will give us just the first few row names, so we will have to define one, and `purrr` gives us a nice shorthand for doing that with `~` syntax.
In this syntax we type a `~` followed by an expression (R code) that uses `.x` as a placeholder for a single element of the list.
`purrr::map()` will then apply the expression to each element of the list, and return the results as a new list.

```{r head_markernames, live = TRUE}
# Get the first few row names of each table with a purrr function.
purrr::map(as.list(markers), # convert markers to a 'regular' list for purrr

>            ~ rownames(head(.x)) )#  our custom function.

Reviewer 2: ```suggestion
           ~ rownames(head(.x))) # our custom function.
```

```

Another variant is `purrr::imap()`, which allows us to use the names of the list elements in our function. 

> (UTry`names(markers)` to see the names for the list we are working with now.)

Reviewer 2: ```suggestion
(Try `names(markers)` to see the names for the list we are working with now.)
```

We will use that here to name output files where we will print each of the marker tables, one for each cell type.
We are again defining a custom function within the call to `purrr:imap()` using the `~` syntax, but this time we can use `.x` for the list elements and `.y` for their names.

In addition to writing the tables out, we are saving the data frames we created as a new list object that we can use in the next step.

```{r write_tables}
marker_df_list <- purrr::imap(
  as.list(markers), # convert markers to a 'regular' list for purrr
  ~ as.data.frame(.x) %>% # convert the table to a data frame
    tibble::rownames_to_column("gene") %>% # make genes a new column
    readr::write_tsv( # write each data frame to a file
      file.path(marker_dir, paste0(.y, "_markers.tsv")) # construct the output path

>       )

Reviewer 2: ```suggestion
     )
```
I think?

Author: this one looks right to me too?

Author: Never mind, you are correct, I was a tab too far. 📏
